From patchwork Tue Sep  6 19:49:05 2022
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Liam Howlett <Liam.Howlett@Oracle.com>
X-Patchwork-Id: 12968117
Return-Path: <owner-linux-mm@kvack.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
Received: from kanga.kvack.org (kanga.kvack.org [205.233.56.17])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 0F928C54EE9
	for <linux-mm@archiver.kernel.org>; Tue,  6 Sep 2022 19:50:54 +0000 (UTC)
Received: by kanga.kvack.org (Postfix)
	id E1F548D000E; Tue,  6 Sep 2022 15:50:36 -0400 (EDT)
Received: by kanga.kvack.org (Postfix, from userid 40)
	id DD08980014; Tue,  6 Sep 2022 15:50:36 -0400 (EDT)
X-Delivered-To: int-list-linux-mm@kvack.org
Received: by kanga.kvack.org (Postfix, from userid 63042)
	id B373D8D0010; Tue,  6 Sep 2022 15:50:36 -0400 (EDT)
X-Delivered-To: linux-mm@kvack.org
Received: from relay.hostedemail.com (smtprelay0017.hostedemail.com
 [216.40.44.17])
	by kanga.kvack.org (Postfix) with ESMTP id 98D938D000E
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 15:50:36 -0400 (EDT)
Received: from smtpin12.hostedemail.com (a10.router.float.18 [10.200.18.1])
	by unirelay04.hostedemail.com (Postfix) with ESMTP id 678841A0473
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:50:36 +0000 (UTC)
X-FDA: 79882702872.12.77CF56A
Received: from mx0a-00069f02.pphosted.com (mx0a-00069f02.pphosted.com
 [205.220.165.32])
	by imf21.hostedemail.com (Postfix) with ESMTP id 03E601C0081
	for <linux-mm@kvack.org>; Tue,  6 Sep 2022 19:50:35 +0000 (UTC)
Received: from pps.filterd (m0246629.ppops.net [127.0.0.1])
	by mx0b-00069f02.pphosted.com (8.17.1.5/8.17.1.5) with ESMTP id
 286IdAjA024665;
	Tue, 6 Sep 2022 19:50:26 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : references : in-reply-to : content-type :
 content-transfer-encoding : mime-version; s=corp-2022-7-12;
 bh=Ofq0r9tv8WG6MaH/Kzgbm8soGh96DPJQwoND4sCdjI8=;
 b=1MG+MQDyp7Q5ETOZ88cFSSoryZkDV2acoVcI2f5P6OtjhcfumCJRrVw90G8rD7FQSmsa
 zTHR/paGaHrcLwagxKVbws7ZJcArPm0BqxhibZd5N4R0EhpWAhCGijcp9nxqLduQSxOl
 hutHPsgJqio2Sddr1v5RpdOnKViDlRTSAOYwJjg1uhtAuKyaB75GJFO8rxiWIHmcJT4T
 0hdmRRa8yhAHnQx66AvQVAdChQh2w/g6Inpg+qDJ9popfQPVe6q5tnUmAIEkXT0A/Ob4
 EJ+fN8G3/cxcjBYx9svyEN4WPDX2fU1mMIzKEIUtsEhjdm8A5jcT7ukiLtWJka5rpDV0 RQ==
Received: from iadpaimrmta02.imrmtpd1.prodappiadaev1.oraclevcn.com
 (iadpaimrmta02.appoci.oracle.com [147.154.18.20])
	by mx0b-00069f02.pphosted.com (PPS) with ESMTPS id 3jbxtaet9q-1
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:50:26 +0000
Received: from pps.filterd
 (iadpaimrmta02.imrmtpd1.prodappiadaev1.oraclevcn.com [127.0.0.1])
	by iadpaimrmta02.imrmtpd1.prodappiadaev1.oraclevcn.com (8.17.1.5/8.17.1.5)
 with ESMTP id 286H8hVY027546;
	Tue, 6 Sep 2022 19:50:24 GMT
Received: from nam10-dm6-obe.outbound.protection.outlook.com
 (mail-dm6nam10lp2105.outbound.protection.outlook.com [104.47.58.105])
	by iadpaimrmta02.imrmtpd1.prodappiadaev1.oraclevcn.com (PPS) with ESMTPS id
 3jbwc9kxfb-3
	(version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256 verify=OK);
	Tue, 06 Sep 2022 19:50:24 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=kJ5sL3scvrUMQbWTkN4sILlCmJ0siem5LeeSsKt1WLbX1joSHVaL/tNUD9Mkm1wgAP6zF7QP/tEGiaP7ZjjI5sCmT13EOG+Ai7GeNEnFRefJoGt+TsOIIjiNqP6ZJX9tYLg42i4DuP7gPl5KDGRo8b9sptpuq/zfQg12aysw6K8kCafe1QzF6xd42ZmqnaUNW6i81JKSkGKI3l7ImHsz8EPCvccR3v+8+lNCC9/dbmizj314oFR02i/ZpzasrivgVS+zTEov081luf1QwOBCvH045MgVdpf0Id/7yHnUMHImdfuyJjqaH9ZVUsoE0UPEoCYdzvHVGOrCcJaNtcK08A==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-AntiSpam-MessageData-ChunkCount:X-MS-Exchange-AntiSpam-MessageData-0:X-MS-Exchange-AntiSpam-MessageData-1;
 bh=Ofq0r9tv8WG6MaH/Kzgbm8soGh96DPJQwoND4sCdjI8=;
 b=e7YGnCBMM5cZMsPRzoTcxSsJ5eikP9XsPJOyH+3mWc56EfiF1H/4u4JP+d1azrDlNFQibPd3Y4LpvhPSM0+WKv+J0ovKuQASQ7Q35wq32n7mJU2dizH5hpIWPLugSPFi8spi7UktcaWDypS6yiP0hWpi9UnGW6J34ELOuE0bGWgKUOmkHNOw0wGON0ndGTyAPvShRdVGpoCrGHTXwPgoZ1F0ltoepWyZHOrKu1CwsATZaqNHVT9IBB9xFVeCxb9So/aYX5tNJTUf3utBCqZeTk9PtOvEfL5GenFmidIPqqys+VG870PcY7m+oGKHl3Tmoy4yx/+aTTVpt5kjNUEQ7A==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Ofq0r9tv8WG6MaH/Kzgbm8soGh96DPJQwoND4sCdjI8=;
 b=DSGgh+PL9orjeX93ImoSdgLQgDWqBM/g8P1m9ex+arRF+bPZl5UnIQPxaPqz56ovTT+CW+OhVzKUNY4SRubFLexOJ8BHldmjShhcyIBRIWj+I1kCazjA3roujONnIOVqbR5iIfuLx3VE+y3oYZX4fGxEGylU7z2xfPqlShn2tzM=
Received: from SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25)
 by SJ0PR10MB5598.namprd10.prod.outlook.com (2603:10b6:a03:3d9::18) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.5588.10; Tue, 6 Sep
 2022 19:50:22 +0000
Received: from SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223]) by SN6PR10MB3022.namprd10.prod.outlook.com
 ([fe80::a420:3107:436d:d223%5]) with mapi id 15.20.5588.018; Tue, 6 Sep 2022
 19:50:22 +0000
From: Liam Howlett <liam.howlett@oracle.com>
To: "maple-tree@lists.infradead.org" <maple-tree@lists.infradead.org>,
        "linux-mm@kvack.org" <linux-mm@kvack.org>,
        "linux-kernel@vger.kernel.org"
	<linux-kernel@vger.kernel.org>,
        Andrew Morton <akpm@linux-foundation.org>
CC: "Matthew Wilcox (Oracle)" <willy@infradead.org>,
        Yang Yingliang
	<yangyingliang@huawei.com>,
        Liam Howlett <liam.howlett@oracle.com>,
        Vlastimil
 Babka <vbabka@suse.cz>
Subject: [PATCH v14 65/70] nommu: remove uses of VMA linked list
Thread-Topic: [PATCH v14 65/70] nommu: remove uses of VMA linked list
Thread-Index: AQHYwim4T8c+LB+VU0+AKrzYG957dQ==
Date: Tue, 6 Sep 2022 19:49:05 +0000
Message-ID: <20220906194824.2110408-66-Liam.Howlett@oracle.com>
References: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
In-Reply-To: <20220906194824.2110408-1-Liam.Howlett@oracle.com>
Accept-Language: en-US
Content-Language: en-US
X-MS-Has-Attach: 
X-MS-TNEF-Correlator: 
x-mailer: git-send-email 2.35.1
x-ms-publictraffictype: Email
x-ms-office365-filtering-correlation-id: 4cf69de8-7c39-4500-f027-08da90410905
x-ms-traffictypediagnostic: SJ0PR10MB5598:EE_
x-ms-exchange-senderadcheck: 1
x-ms-exchange-antispam-relay: 0
x-microsoft-antispam: BCL:0;
x-microsoft-antispam-message-info: 
 5mabmq1LptqDX1f/mV1R5PDI76jQc88jaugoiilHq2EYVa/4R9USbeG5AvDuu2Uw6z0H4GEsrcrkTKUx70kgZsBveilg+wlbQd3XOonfu6mwnFlPHwVMqUCyIx1szINrYNawvfpPqBy9WCMRUD6kMK8hpKofbGHW0jy/I3coQClrVliHeXeMNEfPsz6yLJf2grL2+wmEyV0aSB7EekYu32VwT7wEmTNENQaOUiYdZGJTQFJu8sQ/RgkV1FyswDQUTETz48XlcJeUGOrCnvCruPZmJxLyxmT48m9yNWwKxwMtGxWbq3Z+ZlN5nuzbh36mHZyAU9hUTqAKcvijL6RnT7PpibglImyAPjyFdmiGL6q8NDJD5YjdJ3IorF0XhbeG1ln1KytIJCdrYrC6504+1UxLxbqClvFiYCJOwHfwrCvf4dcD/nnkihRfZgDkKadaKaX4gyh+jbk2ptOwRqCGrv5O2Gf4ypZSJxD0hYq80VUpVmpwfyUhEGc6ZErSO4nIzZzI3CF5wUBWP3PT8D54q0FZOcCaRF/J4BMY6QJo0bHNtDhxv5YcoynFoZ5dLF+4Z0/N1ikxj+KTvYyCCQG1bZ0FyHNjK85JgItiHoagCwVatSwdnWv5RVN/aPRYYQtVEp7a4lPPx/qOmGZoRKLDpn8LfOAYqTHr8xILRsHvsf21nRVm/DzhvVjmgC/LHLdaRrIU3ofRVw3nXVr1vbdSk+/6fGr2sKZNxt2abVA0xjvS3ghASrL1ue1rGlNcL4VPQq4GuMPU8sMupMMVohPKgQ==
x-forefront-antispam-report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3022.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(13230016)(346002)(39860400002)(396003)(136003)(376002)(366004)(64756008)(66556008)(41300700001)(91956017)(76116006)(66476007)(66446008)(66946007)(71200400001)(54906003)(110136005)(36756003)(316002)(6512007)(44832011)(8676002)(122000001)(2906002)(6486002)(4326008)(186003)(8936002)(5660300002)(478600001)(38070700005)(38100700002)(2616005)(86362001)(26005)(1076003)(6666004)(6506007)(83380400001);DIR:OUT;SFP:1101;
x-ms-exchange-antispam-messagedata-chunkcount: 1
x-ms-exchange-antispam-messagedata-0: =?iso-8859-1?q?NEqj8+yMm+DGoMiFIGTIZEs?=
	=?iso-8859-1?q?LiViyDf0Ief1WNmam2mTCeMAGt3QKeeYVbOVm2/vkOI5EpM4GqU7U7cIU+II?=
	=?iso-8859-1?q?xZa3szIep2xKBnDGeTODLcx3Gdq/J9frmqZ9LExkufRnTpjMs3QvwkWpcPpt?=
	=?iso-8859-1?q?H4X00MN/jAuCBLQUOStiDxxuQMBreKIiKvu8aSutfEd3InyrAwJWwG0pS/Au?=
	=?iso-8859-1?q?fSOACCnYo5ez2quD8+Nsj1YLHkAKKSwVkwczC6V9GZJ+LxlDejH5JxmxR/5H?=
	=?iso-8859-1?q?71+ZkrnNOOuio3hJk442aw5eWsYToGa0oy3a4bezGAgbcmj7FWOigvtpsm13?=
	=?iso-8859-1?q?Y7QhNPftNW3Pi/7Q8Q8HB3Eu98ePCjxMlNqT+qARHiUFCtYLAa/Lh0+U94Jk?=
	=?iso-8859-1?q?n6MPKHMx4VMCV7hPjsid2OjJkDYfeBrYNq58Gdbo8N6XpYReg0XaVKrcE7OR?=
	=?iso-8859-1?q?0jm8NdKfu2cCMRhuJc71a2jwxjagbs9PYRqQo/KbqJJ39pbLe8keeusJlEgn?=
	=?iso-8859-1?q?Z6Zg3yi/rdNVgDD/i73YqcH9WduYKmJheeTd7t9sJQsIIb7kxeK1ZK/pHkFj?=
	=?iso-8859-1?q?LM7o3Z+OShxVwgq4ENq2PqvlJJTr7DODPgNtnymQZeXq3ixGPcyVQL63WT0h?=
	=?iso-8859-1?q?x4wQK3xvkoBHA54nBiR5oUJPjVvskejcjToX+9IbJrRijmQbqrUN4hV5gIfu?=
	=?iso-8859-1?q?OrsHUHql8hWD7TOMS7jAZ8eLap+ydqv9uCCmRE0KrQ/I3qRwY8E2ILSvbruW?=
	=?iso-8859-1?q?+pLf1l22f4H8gO1bup5+bHMBOBStS9AJp0Cw8wH6OM94YBbit9R9YmkOUKQi?=
	=?iso-8859-1?q?NbcP0/POXn39Fdhr0IHOpPn9tt2rPKAC6RaWVr7LgRVnQRanHzfvEaXf/2xm?=
	=?iso-8859-1?q?2sq1wgqBzP1yz0Rnr0Ufb0l+Pa7xBnjW/9IkrfC+AYJfr2+JppjB1AJ8H7yz?=
	=?iso-8859-1?q?+YnoUxc8ZfkBaHr5kF8l2td2Q6+d+jp5NMbc3Ql53hZlmVdjoe0lXh3rIcTO?=
	=?iso-8859-1?q?rjMQ1jkUDZrzX9/r4MbHUJw3/1nhhBglLyhlbmc2zK5q3mHFS1mLPZISgtlV?=
	=?iso-8859-1?q?tbwssfyz6uxXxBUe8AHaehGLCPoUnEJDRm9vW0hTfqjErFAU2sChQvb2dG72?=
	=?iso-8859-1?q?TMF+DuQUAwdkAcarrkfn9uJVd8TjalwdeZtMJptc9imZ9qczABV671PLW9gA?=
	=?iso-8859-1?q?KiceRzo4ubDdbjNX9ic+AmgraOUHR82J/i47Sy2mcHBUZA7vNhf75+DxXJcW?=
	=?iso-8859-1?q?DzbgiN0b/uLXCgz/78D1QbpKENMooReH3NlOSSPAvyu8Kv+uIjGzaG4vzmkH?=
	=?iso-8859-1?q?HOamtgPuQR1KLquDtqlI1yRvfz3CrcS9a+ENXIE3BtFTS6L6G1rqnnYDvz/B?=
	=?iso-8859-1?q?LVcUcz06lhKanLfbw9JClKigj/lKPiJLt+qK4SgdcV1YAnNjeNZEMvVA4QiT?=
	=?iso-8859-1?q?qNrZiMEvN8NtVW72GjndDu5KoCWXP0LcaW0B8WBsSuIfMaSVbeG8SXxhrO0X?=
	=?iso-8859-1?q?mre73vJLAbYE6X75f4QE5emuWQGxtBw15dV2A7aVVgvfumTTIFaluq+cCypF?=
	=?iso-8859-1?q?5RHedFcmYsa5YlME6oJBb3q+tIVZ6a8Oz+Qb/1Ye+oUeUy3DrEir0Gttr26e?=
	=?iso-8859-1?q?Hu4j0+RnyrlaUqtxlaVaUHb2ncrvenqdMmaqRjA=3D=3D?=
MIME-Version: 1.0
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3022.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 4cf69de8-7c39-4500-f027-08da90410905
X-MS-Exchange-CrossTenant-originalarrivaltime: 06 Sep 2022 19:49:05.1307
 (UTC)
X-MS-Exchange-CrossTenant-fromentityheader: Hosted
X-MS-Exchange-CrossTenant-id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-mailboxtype: HOSTED
X-MS-Exchange-CrossTenant-userprincipalname: 
 xvG/VZYVq+7Hm93bRckxt97UmjLs1nohCF7Mw7HDD+H0ymsr9dPYi9OKf2IfHHSsB6UD5N7483llA4L2Nzcugg==
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SJ0PR10MB5598
X-Proofpoint-Virus-Version: vendor=baseguard
 engine=ICAP:2.0.205,Aquarius:18.0.895,Hydra:6.0.528,FMLib:17.11.122.1
 definitions=2022-09-06_09,2022-09-06_02,2022-06-22_01
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 adultscore=0
 phishscore=0 spamscore=0
 mlxlogscore=999 mlxscore=0 bulkscore=0 malwarescore=0 suspectscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2207270000
 definitions=main-2209060091
X-Proofpoint-ORIG-GUID: hKhWdmRn4dOckKE0A4BCX4Oe0nvwPSYc
X-Proofpoint-GUID: hKhWdmRn4dOckKE0A4BCX4Oe0nvwPSYc
ARC-Seal: i=2; s=arc-20220608; d=hostedemail.com; t=1662493836; a=rsa-sha256;
	cv=pass;
	b=OkU6r1prRv8c//0xQG6shg+EJdfTtR6N2T/+otMfwMq83WDdu6vPWL5jeJnAGc95p1CT0W
	0CxuZcvLu54SY3CYwR+2lmNq8wgcg4wd4eA7kFw51AnRXUW8iuZ7GEy/7tk5hlveS1AFq2
	z+sbEPlphhZQ6WJpX55qAkpA0Y5MtuQ=
ARC-Authentication-Results: i=2;
	imf21.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=1MG+MQDy;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=DSGgh+PL;
	spf=pass (imf21.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
ARC-Message-Signature: i=2; a=rsa-sha256; c=relaxed/relaxed;
 d=hostedemail.com;
	s=arc-20220608; t=1662493836;
	h=from:from:sender:reply-to:subject:subject:date:date:
	 message-id:message-id:to:to:cc:cc:mime-version:mime-version:
	 content-type:content-type:
	 content-transfer-encoding:content-transfer-encoding:
	 in-reply-to:in-reply-to:references:references:dkim-signature;
	bh=Ofq0r9tv8WG6MaH/Kzgbm8soGh96DPJQwoND4sCdjI8=;
	b=esJdMpaJN+l2Lma4dmNvDPzzst8iEGubV3ajkqJ9xoLZMW1FlsdPeCkQbr3Iid4HZQ2q4P
	hf42y1XJCoQ4DSR/9RO1mAqdmvbRzIv+iCjRFoPv3CjgPr/rITjcSFyuTb5B/Q2Tqfsimu
	3FTkdQ53x8lnc1VxvhTi6btbJYHTkVM=
Authentication-Results: imf21.hostedemail.com;
	dkim=pass header.d=oracle.com header.s=corp-2022-7-12 header.b=1MG+MQDy;
	dkim=pass header.d=oracle.onmicrosoft.com
 header.s=selector2-oracle-onmicrosoft-com header.b=DSGgh+PL;
	spf=pass (imf21.hostedemail.com: domain of liam.howlett@oracle.com designates
 205.220.165.32 as permitted sender) smtp.mailfrom=liam.howlett@oracle.com;
	dmarc=pass (policy=none) header.from=oracle.com;
	arc=pass ("microsoft.com:s=arcselector9901:i=1")
X-Rspam-User: 
X-Rspamd-Server: rspam06
X-Stat-Signature: umuwn1mgshsnpm8shxkczsb8yk4gbr7y
X-Rspamd-Queue-Id: 03E601C0081
X-HE-Tag: 1662493835-247153
X-Bogosity: Ham, tests=bogofilter, spamicity=0.000000, version=1.2.4
Sender: owner-linux-mm@kvack.org
Precedence: bulk
X-Loop: owner-majordomo@kvack.org
List-ID: <linux-mm.kvack.org>

From: "Matthew Wilcox (Oracle)" <willy@infradead.org>

Use the maple tree or VMA iterator instead.  This is faster and will allow
us to shrink the VMA.

Signed-off-by: Matthew Wilcox (Oracle) <willy@infradead.org>
Signed-off-by: Liam R. Howlett <Liam.Howlett@Oracle.com>
Acked-by: Vlastimil Babka <vbabka@suse.cz>
---
 mm/nommu.c | 146 +++++++++++++++++++++++++++++++++++++++--------------
 1 file changed, 109 insertions(+), 37 deletions(-)

diff --git a/mm/nommu.c b/mm/nommu.c
index 265a444a2cc2..269df51e9226 100644
--- a/mm/nommu.c
+++ b/mm/nommu.c
@@ -557,26 +557,14 @@ void vma_mas_remove(struct vm_area_struct *vma, struct ma_state *mas)
 	mas_store_prealloc(mas, NULL);
 }
 
-/*
- * add a VMA into a process's mm_struct in the appropriate place in the list
- * and tree and add to the address space's page tree also if not an anonymous
- * page
- * - should be called with mm->mmap_lock held writelocked
- */
-static void add_vma_to_mm(struct mm_struct *mm, struct vm_area_struct *vma)
+static void setup_vma_to_mm(struct vm_area_struct *vma, struct mm_struct *mm)
 {
-	struct address_space *mapping;
-	struct vm_area_struct *prev;
-	MA_STATE(mas, &mm->mm_mt, vma->vm_start, vma->vm_end);
-
-	BUG_ON(!vma->vm_region);
-
 	mm->map_count++;
 	vma->vm_mm = mm;
 
 	/* add the VMA to the mapping */
 	if (vma->vm_file) {
-		mapping = vma->vm_file->f_mapping;
+		struct address_space *mapping = vma->vm_file->f_mapping;
 
 		i_mmap_lock_write(mapping);
 		flush_dcache_mmap_lock(mapping);
@@ -584,21 +572,52 @@ static void add_vma_to_mm(struct mm_struct *mm, struct vm_area_struct *vma)
 		flush_dcache_mmap_unlock(mapping);
 		i_mmap_unlock_write(mapping);
 	}
+}
 
-	prev = mas_prev(&mas, 0);
-	mas_reset(&mas);
+/*
+ * mas_add_vma_to_mm() - Maple state variant of add_mas_to_mm().
+ * @mas: The maple state with preallocations.
+ * @mm: The mm_struct
+ * @vma: The vma to add
+ *
+ */
+static void mas_add_vma_to_mm(struct ma_state *mas, struct mm_struct *mm,
+			      struct vm_area_struct *vma)
+{
+	struct vm_area_struct *prev;
+
+	BUG_ON(!vma->vm_region);
+
+	setup_vma_to_mm(vma, mm);
+
+	prev = mas_prev(mas, 0);
+	mas_reset(mas);
 	/* add the VMA to the tree */
-	vma_mas_store(vma, &mas);
+	vma_mas_store(vma, mas);
 	__vma_link_list(mm, vma, prev);
 }
 
 /*
- * delete a VMA from its owning mm_struct and address space
+ * add a VMA into a process's mm_struct in the appropriate place in the list
+ * and tree and add to the address space's page tree also if not an anonymous
+ * page
+ * - should be called with mm->mmap_lock held writelocked
  */
-static void delete_vma_from_mm(struct vm_area_struct *vma)
+static int add_vma_to_mm(struct mm_struct *mm, struct vm_area_struct *vma)
 {
-	MA_STATE(mas, &vma->vm_mm->mm_mt, 0, 0);
+	MA_STATE(mas, &mm->mm_mt, vma->vm_start, vma->vm_end);
+
+	if (mas_preallocate(&mas, vma, GFP_KERNEL)) {
+		pr_warn("Allocation of vma tree for process %d failed\n",
+		       current->pid);
+		return -ENOMEM;
+	}
+	mas_add_vma_to_mm(&mas, mm, vma);
+	return 0;
+}
 
+static void cleanup_vma_from_mm(struct vm_area_struct *vma)
+{
 	vma->vm_mm->map_count--;
 	/* remove the VMA from the mapping */
 	if (vma->vm_file) {
@@ -611,10 +630,25 @@ static void delete_vma_from_mm(struct vm_area_struct *vma)
 		flush_dcache_mmap_unlock(mapping);
 		i_mmap_unlock_write(mapping);
 	}
+}
+/*
+ * delete a VMA from its owning mm_struct and address space
+ */
+static int delete_vma_from_mm(struct vm_area_struct *vma)
+{
+	MA_STATE(mas, &vma->vm_mm->mm_mt, 0, 0);
+
+	if (mas_preallocate(&mas, vma, GFP_KERNEL)) {
+		pr_warn("Allocation of vma tree for process %d failed\n",
+		       current->pid);
+		return -ENOMEM;
+	}
+	cleanup_vma_from_mm(vma);
 
 	/* remove from the MM's tree and list */
 	vma_mas_remove(vma, &mas);
 	__vma_unlink_list(vma->vm_mm, vma);
+	return 0;
 }
 
 /*
@@ -1024,6 +1058,7 @@ unsigned long do_mmap(struct file *file,
 	vm_flags_t vm_flags;
 	unsigned long capabilities, result;
 	int ret;
+	MA_STATE(mas, &current->mm->mm_mt, 0, 0);
 
 	*populate = 0;
 
@@ -1042,6 +1077,7 @@ unsigned long do_mmap(struct file *file,
 	 * now know into VMA flags */
 	vm_flags = determine_vm_flags(file, prot, flags, capabilities);
 
+
 	/* we're going to need to record the mapping */
 	region = kmem_cache_zalloc(vm_region_jar, GFP_KERNEL);
 	if (!region)
@@ -1051,6 +1087,9 @@ unsigned long do_mmap(struct file *file,
 	if (!vma)
 		goto error_getting_vma;
 
+	if (mas_preallocate(&mas, vma, GFP_KERNEL))
+		goto error_maple_preallocate;
+
 	region->vm_usage = 1;
 	region->vm_flags = vm_flags;
 	region->vm_pgoff = pgoff;
@@ -1191,7 +1230,7 @@ unsigned long do_mmap(struct file *file,
 	current->mm->total_vm += len >> PAGE_SHIFT;
 
 share:
-	add_vma_to_mm(current->mm, vma);
+	mas_add_vma_to_mm(&mas, current->mm, vma);
 
 	/* we flush the region from the icache only when the first executable
 	 * mapping of it is made  */
@@ -1217,6 +1256,7 @@ unsigned long do_mmap(struct file *file,
 
 sharing_violation:
 	up_write(&nommu_region_sem);
+	mas_destroy(&mas);
 	pr_warn("Attempt to share mismatched mappings\n");
 	ret = -EINVAL;
 	goto error;
@@ -1233,6 +1273,14 @@ unsigned long do_mmap(struct file *file,
 			len, current->pid);
 	show_free_areas(0, NULL);
 	return -ENOMEM;
+
+error_maple_preallocate:
+	kmem_cache_free(vm_region_jar, region);
+	vm_area_free(vma);
+	pr_warn("Allocation of vma tree for process %d failed\n", current->pid);
+	show_free_areas(0, NULL);
+	return -ENOMEM;
+
 }
 
 unsigned long ksys_mmap_pgoff(unsigned long addr, unsigned long len,
@@ -1298,6 +1346,7 @@ int split_vma(struct mm_struct *mm, struct vm_area_struct *vma,
 	struct vm_area_struct *new;
 	struct vm_region *region;
 	unsigned long npages;
+	MA_STATE(mas, &mm->mm_mt, vma->vm_start, vma->vm_end);
 
 	/* we're only permitted to split anonymous regions (these should have
 	 * only a single usage on the region) */
@@ -1312,9 +1361,13 @@ int split_vma(struct mm_struct *mm, struct vm_area_struct *vma,
 		return -ENOMEM;
 
 	new = vm_area_dup(vma);
-	if (!new) {
-		kmem_cache_free(vm_region_jar, region);
-		return -ENOMEM;
+	if (!new)
+		goto err_vma_dup;
+
+	if (mas_preallocate(&mas, vma, GFP_KERNEL)) {
+		pr_warn("Allocation of vma tree for process %d failed\n",
+			current->pid);
+		goto err_mas_preallocate;
 	}
 
 	/* most fields are the same, copy all, and then fixup */
@@ -1333,7 +1386,6 @@ int split_vma(struct mm_struct *mm, struct vm_area_struct *vma,
 	if (new->vm_ops && new->vm_ops->open)
 		new->vm_ops->open(new);
 
-	delete_vma_from_mm(vma);
 	down_write(&nommu_region_sem);
 	delete_nommu_region(vma->vm_region);
 	if (new_below) {
@@ -1346,9 +1398,19 @@ int split_vma(struct mm_struct *mm, struct vm_area_struct *vma,
 	add_nommu_region(vma->vm_region);
 	add_nommu_region(new->vm_region);
 	up_write(&nommu_region_sem);
-	add_vma_to_mm(mm, vma);
-	add_vma_to_mm(mm, new);
+
+	setup_vma_to_mm(vma, mm);
+	setup_vma_to_mm(new, mm);
+	mas_set_range(&mas, vma->vm_start, vma->vm_end - 1);
+	mas_store(&mas, vma);
+	vma_mas_store(new, &mas);
 	return 0;
+
+err_mas_preallocate:
+	vm_area_free(new);
+err_vma_dup:
+	kmem_cache_free(vm_region_jar, region);
+	return -ENOMEM;
 }
 
 /*
@@ -1363,12 +1425,14 @@ static int shrink_vma(struct mm_struct *mm,
 
 	/* adjust the VMA's pointers, which may reposition it in the MM's tree
 	 * and list */
-	delete_vma_from_mm(vma);
+	if (delete_vma_from_mm(vma))
+		return -ENOMEM;
 	if (from > vma->vm_start)
 		vma->vm_end = from;
 	else
 		vma->vm_start = to;
-	add_vma_to_mm(mm, vma);
+	if (add_vma_to_mm(mm, vma))
+		return -ENOMEM;
 
 	/* cut the backing region down to size */
 	region = vma->vm_region;
@@ -1396,9 +1460,10 @@ static int shrink_vma(struct mm_struct *mm,
  */
 int do_munmap(struct mm_struct *mm, unsigned long start, size_t len, struct list_head *uf)
 {
+	MA_STATE(mas, &mm->mm_mt, start, start);
 	struct vm_area_struct *vma;
 	unsigned long end;
-	int ret;
+	int ret = 0;
 
 	len = PAGE_ALIGN(len);
 	if (len == 0)
@@ -1407,7 +1472,7 @@ int do_munmap(struct mm_struct *mm, unsigned long start, size_t len, struct list
 	end = start + len;
 
 	/* find the first potentially overlapping VMA */
-	vma = find_vma(mm, start);
+	vma = mas_find(&mas, end - 1);
 	if (!vma) {
 		static int limit;
 		if (limit < 5) {
@@ -1426,7 +1491,7 @@ int do_munmap(struct mm_struct *mm, unsigned long start, size_t len, struct list
 				return -EINVAL;
 			if (end == vma->vm_end)
 				goto erase_whole_vma;
-			vma = vma->vm_next;
+			vma = mas_next(&mas, end - 1);
 		} while (vma);
 		return -EINVAL;
 	} else {
@@ -1448,9 +1513,10 @@ int do_munmap(struct mm_struct *mm, unsigned long start, size_t len, struct list
 	}
 
 erase_whole_vma:
-	delete_vma_from_mm(vma);
+	if (delete_vma_from_mm(vma))
+		ret = -ENOMEM;
 	delete_vma(mm, vma);
-	return 0;
+	return ret;
 }
 
 int vm_munmap(unsigned long addr, size_t len)
@@ -1475,6 +1541,7 @@ SYSCALL_DEFINE2(munmap, unsigned long, addr, size_t, len)
  */
 void exit_mmap(struct mm_struct *mm)
 {
+	VMA_ITERATOR(vmi, mm, 0);
 	struct vm_area_struct *vma;
 
 	if (!mm)
@@ -1482,13 +1549,18 @@ void exit_mmap(struct mm_struct *mm)
 
 	mm->total_vm = 0;
 
-	while ((vma = mm->mmap)) {
-		mm->mmap = vma->vm_next;
-		delete_vma_from_mm(vma);
+	/*
+	 * Lock the mm to avoid assert complaining even though this is the only
+	 * user of the mm
+	 */
+	mmap_write_lock(mm);
+	for_each_vma(vmi, vma) {
+		cleanup_vma_from_mm(vma);
 		delete_vma(mm, vma);
 		cond_resched();
 	}
 	__mt_destroy(&mm->mm_mt);
+	mmap_write_unlock(mm);
 }
 
 int vm_brk(unsigned long addr, unsigned long len)

